aips/aip-x.md

```diff
@@ -0,0 +1,247 @@
+    _witness: Witness,
+)
+```
+There are two apis associated with claiming/consuming an on-chain intent. The first API will allow anyone to consume an intent object. Consumption of the intent object can release the resource locked in the intent temporarily. There will also be a hot potato type returned called `TradeSession`. Since this is a hot potato type, the transaction sender will have no choice but invoke the second `finish_intent_session` api which will require a value of `Witness` type. 
+
+### Example: Defining an intent to receive a certain fungible asset.
+```
+module aptos_framework::fungible_asset_intent_hooks {
+    struct FungibleAssetLimitOrder has store, drop {
+        desired_metadata: Object<Metadata>,
+        desired_amount: u64,
+        issuer: address,
+    }
+
+    struct FungibleAssetRecipientWitness has drop {}
+
+    public fun finish_fa_receiving_session(
+        session: TradeSession<FungibleAssetLimitOrder>,
+        received_fa: FungibleAsset,
+    ) {
+        let argument = intent::get_argument(&session);
+        assert!(
+            fungible_asset::metadata_from_asset(&received_fa) == argument.desired_metadata,
+            error::invalid_argument(ENOT_DESIRED_TOKEN)
+        );
+        assert!(
+            fungible_asset::amount(&received_fa) >= argument.desired_amount,
+            error::invalid_argument(EAMOUNT_NOT_MEET),
+        );
+
+        primary_fungible_store::deposit(argument.issuer, received_fa);
+        intent::finish_intent_session(session, FungibleAssetRecipientWitness {})
+    }
+}
+```
+Here we are defining an example of intent where the unlock condition is receiving a certain amount of a given fungible asset type. `Argument` type is instantiated to `FungibleAssetLimitOrder` and `Witness` type is instantated to `FungibleAssetRecipientWitness` in this example. In the arugment, we specify the amount and the type of FA we would like to receive. The check will then be implemented by first checking if the passed in `FungibleAsset` meets the requirement set in the `FungibleAssetLimitOrder` and deposit the FA to the issuer's primary fungible storage if the conditions are met. Once the checks are completed, we will provide the witness `FungibleAssetRecipientWitness` to complete this intent session.
+
+
+## Reference Implementation
+
+
+
+## Testing 
+
+Implemented tests in the framework.
+
+## Risks and Drawbacks
+
+### Frontrun Risk
+
+The intent claiming transaction can be front-runned by malicious mempool operator. Since anyone is able to sign transaction to claim intent solution, a malicous mempool operator can take the intent solving transaction submitted by the solver and sign the same transaction with its own address to be able to claim the lucrative ones. However, since txns usually won't stay too long in the mempool right now, it should not be a major concern at this point. In the future we might need some authentication schema for intent claiming process.
```

Comment:
I would rephrase this:

> As the system is first-come first-serve, the intent claiming transaction can be front-runned by whoever sees it - mempool operator, validator operator, etc. Since anyone is able to sign transaction to claim intent solution, anyone can take the intent solving transaction submitted by the solver and sign the same transaction with its own address to be able to claim the lucrative ones. If main value comes from finding the route, and not being willing to settle, privacy of submitted transactions will become imporatnt. Intent Solver might want to run a validator node or partner with validator operator. Additionally in the future, support for "secret transactions" , that are revealed only at the time of execution might be added.